src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,286 @@
+#include <Parsers/ASTFunction.h>
+#include <Interpreters/ArithmeticOperationsInAgrFuncOptimize.h>
+#include <IO/WriteHelpers.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int UNEXPECTED_AST_STRUCTURE;
+}
+
+bool isConstantField(const Field & field)
+{
+    return field.getType() == Field::Types::Int64 ||
+           field.getType() == Field::Types::UInt64 ||
+           field.getType() == Field::Types::Int128 ||
+           field.getType() == Field::Types::UInt128;
+}
+
+/// scalar values from the first level
+std::pair<ASTs, ASTs> tryGetConst(std::string & name, ASTs & arguments)
+{
+    ASTs const_num;
+    ASTs not_const;
+
+    for (const auto & arg : arguments)
+    {
+        if (const auto * literal = arg->as<ASTLiteral>())
+        {
+            if (isConstantField(literal->value))
+                const_num.push_back(arg);
+            else
+                not_const.push_back(arg);
+        }
+        else
+            not_const.push_back(arg);
+    }
+
+    if ((name == "plus" || name == "multiply") && const_num.size() + not_const.size() != 2)
+    {
+        throw Exception("Wrong number of arguments for function 'plus' or 'multiply' (" + toString(const_num.size() + not_const.size()) + " instead of 2)",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+    }
+
+    return {const_num, not_const};
+}
+
+std::pair<ASTs, ASTs> findAllConsts(ASTFunction * func_node, std::string & inter_func_name)
+{
+    if (!func_node->arguments)
+        return {};
+
+    if (!(func_node->arguments->children[0]->as<ASTFunction>()) && (func_node->arguments->children.size() == 2 &&
+        !(func_node->arguments->children[1]->as<ASTFunction>())))
+    {
+        auto arg = func_node->arguments->children;
+        std::pair<ASTs, ASTs> fl = tryGetConst(func_node->name, arg);
+        return fl;
+    }
+    else if ((func_node->arguments->children[0]->as<ASTFunction>() && inter_func_name != func_node->arguments->children[0]->as<ASTFunction>()->name) ||
+            (func_node->arguments->children.size() == 2 && func_node->arguments->children[1]->as<ASTFunction>() &&
+            inter_func_name != func_node->arguments->children[1]->as<ASTFunction>()->name))
+    {
+        if (func_node->arguments->children[0]->as<ASTFunction>())
+            return {{func_node->arguments->children[1]}, {func_node->arguments->children[0]}};
+        else
+            return {{func_node->arguments->children[0]}, {func_node->arguments->children[1]}};
+    }
+    else
+    {
+        std::pair<ASTs, ASTs> fl = tryGetConst(func_node->name, func_node->arguments->children);
+        if (!fl.second[0]->as<ASTFunction>())
+        {
+            std::pair<ASTs, ASTs> ans = {fl.first, fl.second};
+            return ans;
+        }
+
+        std::pair<ASTs, ASTs> ans = findAllConsts(fl.second[0]->as<ASTFunction>(), inter_func_name);
+
+        if (fl.first.size() == 1)
```

Comment:
There're `ans.first`, `ans.second`, `fl.first` and `fl.second`. These names are very bad. Make references for all `first`s and `second`s.